ovn-org/ovn-kubernetes
# Add support for Topology Aware Hints

**- What this PR does and why is it needed**
This PR implements the upstream K8s feature: Topology Aware Hints: https://kubernetes.io/docs/concepts/services-networking/topology-aware-hints/ 

**- Special notes for reviewers**
I plan to add an e2e into K8s for inter-interoperability between ITP&ETP&TAH which seems to have no coverage when used together.

**- How to verify it**
The non-HA lane runs the TAH e2e test when using multi-zone environments:
```
2022-09-24T12:16:59.2669822Z [BeforeEach] [sig-network] [Feature:Topology Hints]
2022-09-24T12:16:59.2670197Z   test/e2e/framework/framework.go:187
2022-09-24T12:16:59.2671148Z [1mSTEP[0m: Creating a kubernetes client
2022-09-24T12:16:59.2671491Z Sep 24 12:16:25.806: INFO: >>> kubeConfig: /home/runner/ovn.conf
2022-09-24T12:16:59.2671933Z [1mSTEP[0m: Building a namespace api object, basename topology-hints
2022-09-24T12:16:59.2672400Z [1mSTEP[0m: Waiting for a default service account to be provisioned in namespace
2022-09-24T12:16:59.2672844Z [1mSTEP[0m: Waiting for kube-root-ca.crt to be provisioned in namespace
2022-09-24T12:16:59.2673359Z [BeforeEach] [sig-network] [Feature:Topology Hints]
2022-09-24T12:16:59.2673702Z   test/e2e/network/topology_hints.go:50
2022-09-24T12:16:59.2673971Z [It] should distribute endpoints evenly
2022-09-24T12:16:59.2674251Z   test/e2e/network/topology_hints.go:55
2022-09-24T12:16:59.2675186Z Sep 24 12:16:30.896: INFO: DaemonSet pods can't tolerate node ovn-control-plane with taints [{Key:node-role.kubernetes.io/master Value: Effect:NoSchedule TimeAdded:<nil>} {Key:node-role.kubernetes.io/control-plane Value: Effect:NoSchedule TimeAdded:<nil>}], skip checking this node
2022-09-24T12:16:59.2676023Z Sep 24 12:16:30.926: INFO: Number of nodes with available pods controlled by daemonset topology-serve-hostname: 1
2022-09-24T12:16:59.2676525Z Sep 24 12:16:30.926: INFO: Node ovn-worker is running 0 daemon pod, expected 1
2022-09-24T12:16:59.2677466Z Sep 24 12:16:36.033: INFO: DaemonSet pods can't tolerate node ovn-control-plane with taints [{Key:node-role.kubernetes.io/master Value: Effect:NoSchedule TimeAdded:<nil>} {Key:node-role.kubernetes.io/control-plane Value: Effect:NoSchedule TimeAdded:<nil>}], skip checking this node
2022-09-24T12:16:59.2678206Z Sep 24 12:16:36.056: INFO: Number of nodes with available pods controlled by daemonset topology-serve-hostname: 2
2022-09-24T12:16:59.2679184Z Sep 24 12:16:36.056: INFO: Number of running nodes: 2, number of available pods: 2 in daemonset topology-serve-hostname
2022-09-24T12:16:59.2680092Z Sep 24 12:16:36.081: INFO: DaemonSet pods can't tolerate node ovn-control-plane with taints [{Key:node-role.kubernetes.io/master Value: Effect:NoSchedule TimeAdded:<nil>} {Key:node-role.kubernetes.io/control-plane Value: Effect:NoSchedule TimeAdded:<nil>}], skip checking this node
2022-09-24T12:16:59.2680779Z Sep 24 12:16:36.161: INFO: Waiting for 2 endpoints to be tracked in EndpointSlices
2022-09-24T12:16:59.2681156Z [1mSTEP[0m: having hints set for each endpoint
2022-09-24T12:16:59.2681557Z [1mSTEP[0m: keeping requests in the same zone
2022-09-24T12:16:59.2681947Z [1mSTEP[0m: creating a client pod for probing the service from zone-1
2022-09-24T12:16:59.2682483Z Sep 24 12:16:41.182: INFO: The status of Pod curl-from-zone-1 is Pending, waiting for it to be Running (with Ready = true)
2022-09-24T12:16:59.2683269Z Sep 24 12:16:43.185: INFO: The status of Pod curl-from-zone-1 is Pending, waiting for it to be Running (with Ready = true)
2022-09-24T12:16:59.2683769Z Sep 24 12:16:45.191: INFO: The status of Pod curl-from-zone-1 is Running (Ready = true)
2022-09-24T12:16:59.2684305Z Sep 24 12:16:45.201: INFO: Ensuring that requests from curl-from-zone-1 pod on ovn-worker2 node stay in zone-1 zone
2022-09-24T12:16:59.2684675Z Sep 24 12:16:50.229: INFO: Pod client logs: Sat Sep 24 12:16:43 UTC 2022
2022-09-24T12:16:59.2685027Z Date: Sat Sep 24 12:16:44 UTC 2022 Try: 1
2022-09-24T12:16:59.2685340Z topology-serve-hostname-xcc4c
2022-09-24T12:16:59.2685597Z Date: Sat Sep 24 12:16:46 UTC 2022 Try: 2
2022-09-24T12:16:59.2685880Z topology-serve-hostname-xcc4c
2022-09-24T12:16:59.2686132Z Date: Sat Sep 24 12:16:47 UTC 2022 Try: 3
2022-09-24T12:16:59.2686507Z topology-serve-hostname-xcc4c
2022-09-24T12:16:59.2686750Z Date: Sat Sep 24 12:16:48 UTC 2022 Try: 4
2022-09-24T12:16:59.2687039Z topology-serve-hostname-xcc4c
2022-09-24T12:16:59.2687285Z Date: Sat Sep 24 12:16:49 UTC 2022 Try: 5
2022-09-24T12:16:59.2687560Z topology-serve-hostname-xcc4c
2022-09-24T12:16:59.2687870Z 
2022-09-24T12:16:59.2688118Z [1mSTEP[0m: creating a client pod for probing the service from zone-0
2022-09-24T12:16:59.2688656Z Sep 24 12:16:50.245: INFO: The status of Pod curl-from-zone-0 is Pending, waiting for it to be Running (with Ready = true)
2022-09-24T12:16:59.2689235Z Sep 24 12:16:52.302: INFO: The status of Pod curl-from-zone-0 is Pending, waiting for it to be Running (with Ready = true)
2022-09-24T12:16:59.2689713Z Sep 24 12:16:54.247: INFO: The status of Pod curl-from-zone-0 is Running (Ready = true)
2022-09-24T12:16:59.2690250Z Sep 24 12:16:54.249: INFO: Ensuring that requests from curl-from-zone-0 pod on ovn-worker node stay in zone-0 zone
2022-09-24T12:16:59.2690638Z Sep 24 12:16:59.255: INFO: Pod client logs: Sat Sep 24 12:16:52 UTC 2022
2022-09-24T12:16:59.2690910Z Date: Sat Sep 24 12:16:53 UTC 2022 Try: 1
2022-09-24T12:16:59.2691209Z topology-serve-hostname-zfgqq
2022-09-24T12:16:59.2691459Z Date: Sat Sep 24 12:16:55 UTC 2022 Try: 2
2022-09-24T12:16:59.2691740Z topology-serve-hostname-zfgqq
2022-09-24T12:16:59.2691981Z Date: Sat Sep 24 12:16:56 UTC 2022 Try: 3
2022-09-24T12:16:59.2692275Z topology-serve-hostname-zfgqq
2022-09-24T12:16:59.2692528Z Date: Sat Sep 24 12:16:57 UTC 2022 Try: 4
2022-09-24T12:16:59.2692799Z topology-serve-hostname-zfgqq
2022-09-24T12:16:59.2693034Z Date: Sat Sep 24 12:16:58 UTC 2022 Try: 5
2022-09-24T12:16:59.2693311Z topology-serve-hostname-zfgqq
2022-09-24T12:16:59.2693876Z 
2022-09-24T12:16:59.2694104Z [AfterEach] [sig-network] [Feature:Topology Hints]
2022-09-24T12:16:59.2714990Z   test/e2e/framework/framework.go:188
2022-09-24T12:16:59.2715523Z Sep 24 12:16:59.255: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
2022-09-24T12:16:59.2716229Z [1mSTEP[0m: Destroying namespace "topology-hints-5855" for this suite.
2022-09-24T12:16:59.2716434Z 
2022-09-24T12:16:59.2716441Z 
2022-09-24T12:16:59.2716659Z [32m• [SLOW TEST:33.455 seconds][0m
2022-09-24T12:16:59.2716965Z [sig-network] [Feature:Topology Hints]
2022-09-24T12:16:59.2717322Z [90mtest/e2e/network/common/framework.go:23[0m
2022-09-24T12:16:59.2717609Z   should distribute endpoints evenly
2022-09-24T12:16:59.2717981Z   [90mtest/e2e/network/topology_hints.go:55[0m
2022-09-24T12:16:59.2718299Z [90m------------------------------[0m
2022-09-24T12:16:59.2718830Z {"msg":"PASSED [sig-network] [Feature:Topology Hints] should distribute endpoints evenly",
```
Sufficient Unit tests have been added to test the LB filtering which is the main logic for this feature.

Directly tested on a KIND cluster as well.

**- Description for the changelog**
`Support Topology Aware Hints in OVNK`
